docs/architecture/architecture.md

```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+can be very challenging.
+
+By default the Clear Containers runtime creates a virtual machine for each container,
+but with Kubernetes it will create and launch one virtual machine per pod. Then each
+container within a pod is simply a container inside the VM boundary.
+
+So the challenge with Clear Containers when trying to work as a `CRI-O` runtime is
+to know when to create a full virtual machine (for pods) and when to create a new
+container inside a previously created VM. In both cases it will get called with very
+similar arguments, so it needs `CRI-O` help to be able to distinguish between a pod
+creation from a container one.
+
+### OCI annotations
+
+In order for the Clear Containers runtime (Or any VM based OCI compatible runtime) to
+be able to understand if it needs to create a full virtual machine or if it has to create
+a new container inside an existing pod's VM context, `CRI-O` adds specific annotations
+to the OCI configuration file (`config.json`) passed down to the OCI compatible runtimes
+it relies on.
+
+Before calling its runtime, `CRI-O` will always add a `io.kubernetes.cri-o.ContainerType`
+annotation to the `config.json` OCI configuration files it produces from the kubelet CRI
+request. `io.kubernetes.cri-o.ContainerType` can either be set to `sandbox` or `container`.
+Clear Containers will then use this annotation to decide if it needs to respectively
+create a virtual machine or a container inside a virtual machine associated with a
+Kubernetes pod:
+
+```Go
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return err
+	}
+
+	switch containerType {
+	case vc.PodSandbox:
+		process, err = createPod(ociSpec, runtimeConfig, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	case vc.PodContainer:
+		process, err = createContainer(ociSpec, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	}
+
+```
+
+### Generalization
+
+With Clear Containers 2.1, we implemented the initial Kubernetes/`CRI-O` support by choosing
+to handle pods as a special case for containers. By default the runtime would only be aware
+of containers and only when receiving CRI-O specific annotations it would follow specific
+code paths to handle pods and intra pod containers. This approach showed that container
+orchestration support was an after thought with 2.1, and that supporting the single container
+Docker use case was the main focus.
+
+Starting with Clear Containers 3.0 we took a different, almost opposite approach. The Clear
+Containers 3.0 runtime is based on the `virtcontainers` package which API is inspired by the
+Kubernetes `CRI` one. `virtcontainers` execution units are pods and it does not create or manage
+any container outside of an existing pod context, and thus so does the Clear Container 3.0 runtime.
+To create a Docker container it will first create a pod which will represent the container
```

Comment:
To me it reads better if you reorder the words and refer to "the hardware virtualized container context".